products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,168 @@
+      }
+    }
+    
+    virtual const KeyID getKeyID(const KeyType& key) override {
+      std::lock_guard<std::mutex> guard(keyTableMutex);
+      
+      // The RHS of the mapping is actually ignored, we use the StringMap's ptr
+      // identity because it allows us to efficiently map back to the key string
+      // in `getRuleInfoForKey`.
+      auto it = keyTable.insert(std::make_pair(key, 0)).first;
+      return (KeyID)(uintptr_t)it->getKey().data();
+    }
+    
+    virtual KeyType getKeyForID(const KeyID key) override {
+      // Note that we don't need to lock `keyTable` here because the key entries
+      // themselves don't change once created.
+      return llvm::StringMapEntry<KeyID>::GetStringMapEntryFromKeyData(
+                                                                       (const char*)(uintptr_t)key).getKey();
+    }
+    
+    const bool buildStarted(std::string *error_out) {
+      return _db.get()->buildStarted(error_out);
+    }
+    
+    void buildComplete() {
+      _db.get()->buildComplete();
+    }
+    
+    const bool getKeys(std::vector<KeyType>& keys_out, std::string *error_out) {
+      return _db.get()->getKeys(keys_out, error_out);
+    }
+  };
+  
+}
+
+const llb_data_t mapData(std::vector<uint8_t> input) {
+  const auto data = input.data();
+  const auto size = sizeof(data) * input.size();
+  const auto newData = (uint8_t *)malloc(size);
+  memcpy(newData, data, size);
+  
+  return llb_data_t {
+    input.size(),
+    newData
+  };
+}
+
+const llb_database_t* llb_database_create(
+                                          char *path,
+                                          uint32_t clientSchemaVersion,
+                                          llb_data_t *error_out) {
+  std::string error;
+  
+  auto database = new CAPIBuildDB(StringRef(path), clientSchemaVersion, &error);
+  
+  if (!error.empty() && error_out) {
+    error_out->length = error.size();
+    error_out->data = (const uint8_t*)strdup(error.c_str());
+  }
+  
+  return (llb_database_t *)database;
+}
+
+void llb_database_destroy(llb_database_t *database) {
+  auto db = (CAPIBuildDB *)database;
+  db->buildComplete();
+  delete db;
+}
+
+const llb_database_key_id llb_database_result_keys_get_count(llb_database_result_keys_t *result) {
+  auto resultKeys = (CAPIBuildDBKeysResult *)result;
+  return resultKeys->size();
+}
+
+void llb_database_result_keys_get_key_at_index(llb_database_result_keys_t *result, llb_database_key_id keyID, llb_data_t *key_out) {
+  auto resultKeys = (CAPIBuildDBKeysResult *)result;
+  auto key = resultKeys->keyAtIndex(keyID);
+  key_out->length = key.size();
+  key_out->data = (const uint8_t*) strdup(key.data());
+}
+
+void llb_database_destroy_result_keys(llb_database_result_keys_t *result) {
+  delete (CAPIBuildDBKeysResult *)result;
+}
+
+const bool llb_database_get_keys(llb_database_t *database, llb_database_result_keys_t **keysResult_out, llb_data_t *error_out) {
+  auto db = (CAPIBuildDB *)database;
```

Comment:
For safety, please init `*keysResult_out` to nullptr at start.

Reply:
(Except, as above, I think we should change the signature so this problem will go away)